rust-runtime/smithy-eventstream/src/frame.rs

```diff
@@ -581,10 +556,141 @@ mod tests {
+    /// to be called over and over again with more data in the buffer each time its called.
+    /// When there's not enough data to decode a message, it returns `Ok(None)`.
+    ///
+    /// Once there is enough data to read a message prelude, then it will mutate the `Buf`
+    /// position. The state from the reading of the prelude is stored in the decoder so that
+    /// the next call will be able to decode the entire message, even though the prelude
+    /// is no longer available in the `Buf`.
+    pub fn decode_frame<B: Buf>(&mut self, mut buffer: B) -> Result<DecodedFrame, Error> {
+        if !self.prelude_read && buffer.remaining() >= PRELUDE_LENGTH_BYTES_USIZE {
+            buffer.copy_to_slice(&mut self.prelude);
+            self.prelude_read = true;
+        }
+
+        if let Some(remaining_len) = self.remaining_ready(&buffer)? {
+            let mut message_buf = (&self.prelude[..]).chain(buffer.take(remaining_len));
+            let result = Message::read_from(&mut message_buf).map(DecodedFrame::Complete);
+            self.reset();
+            return result;
+        }
+
+        Ok(DecodedFrame::Incomplete)
+    }
+}
+
+#[cfg(test)]
+mod message_frame_decoder_tests {
+    use super::{DecodedFrame, MessageFrameDecoder};
+    use crate::frame::Message;
+    use bytes::Bytes;
+    use bytes_utils::SegmentedBuf;
+
+    #[test]
+    fn single_streaming_message() {
+        let message = include_bytes!("../test_data/valid_with_all_headers_and_payload");
+
+        let mut decoder = MessageFrameDecoder::new();
+        let mut segmented = SegmentedBuf::new();
+        for i in 0..(message.len() - 1) {
+            segmented.push(&message[i..(i + 1)]);
+            if let DecodedFrame::Complete(_) = decoder.decode_frame(&mut segmented).unwrap() {
+                panic!("incomplete frame shouldn't result in message");
+            }
+        }
+
+        segmented.push(&message[(message.len() - 1)..]);
+        match decoder.decode_frame(&mut segmented).unwrap() {
+            DecodedFrame::Incomplete => panic!("frame should be complete now"),
+            DecodedFrame::Complete(actual) => {
+                let expected = Message::read_from(&mut Bytes::from_static(message)).unwrap();
+                assert_eq!(expected, actual);
+            }
+        }
+    }
+
+    #[test]
+    fn multiple_streaming_messages() {
+        let message1 = include_bytes!("../test_data/valid_with_all_headers_and_payload");
+        let message2 = include_bytes!("../test_data/valid_empty_payload");
+        let message3 = include_bytes!("../test_data/valid_no_headers");
+        let mut repeated = message1.to_vec();
+        repeated.extend_from_slice(message2);
+        repeated.extend_from_slice(message3);
+
+        let mut decoder = MessageFrameDecoder::new();
+        let mut segmented = SegmentedBuf::new();
+        let mut decoded = Vec::new();
+        for window in repeated.chunks(3) {
```

Comment:
slight preference for one byte at a time to make sure that it works with only a single byte

Reply:
could also just repeat this test with a range of window sizes